src/ints/bios_pci.cpp

```diff
@@ -0,0 +1,265 @@
+{
+	static bool already_warned = false;
+	if (!already_warned) {
+		LOG_WARNING("INT1A: PCI function called despite no PCI present");
+		already_warned = true;
+	}
+}
+
+void INT1AB1_Handler()
+{
+	auto set_return_code = [](const ReturnCode code) {
+		reg_ah = enum_val(code);
+		CALLBACK_SCF(code != ReturnCode::Successful);
+	};
+
+	auto select_read_write_address = []() {
+		const auto address = 0x80000000 |
+	                             static_cast<uint32_t>(reg_bx << 8) |
+	                             static_cast<uint32_t>(reg_di & 0xfc);
+		IO_WriteD(port_num_pci_config_address,
+		          static_cast<uint32_t>(address));
+	};
+
+	if (!PCI_IsInitialized()) {
+		// No PCI subsystem
+
+		if (reg_al != 0x01) {
+			warn_no_pci_present();
+		}
+
+		set_return_code(ReturnCode::FuncNotSupported);
+		return;
+	}
+
+	switch (reg_al) {
+	case 0x01: // PCI BIOS Present
+		reg_bx  = 0x0210;     // version 2.10
+		reg_cx  = 0x0000;     // only one PCI bus
+		reg_edx = 0x20494350; // "PCI "
+		reg_edi = PCI_GetPModeInterface();
+		// AL informs which mechanisms are supported:
+		// bit 0: configuration mechanism #1
+		// bit 1: configuration mechanism #2
+		// (either bit 1 or 2 needs to be set)
+		// bit 4: special cycle generation mechanism #1
+		// bit 5: special cycle generation mechanism #2
+		// (either bit 4 or 5 can be set, must match the supported
+		// configuration mechanism)
+		reg_al = 0x01;
+		set_return_code(ReturnCode::Successful);
+		return;
+	case 0x02: // Find PCI Device
+	{
+		// Check if vendor ID is valid
+		if (reg_dx == UINT16_MAX) {
+			set_return_code(ReturnCode::BadVendorId);
+			return;
+		}
+
+		const auto tmp = (reg_cx << 16) | reg_dx;
+		const uint32_t device_tag = static_cast<uint32_t>(tmp);
+
+		// Try to find requested device
+		uint32_t device_number = 0;
+		for (uint16_t i = 0; i <= max_device_index; ++i) {
+			// Query unique device/subdevice entries
+			const auto i_shifted = static_cast<uint32_t>(i << 8);
+			IO_WriteD(port_num_pci_config_address,
+			          static_cast<uint32_t>(0x80000000 | i_shifted));
+			if (IO_ReadD(0xcfc) != device_tag) {
+				continue; // Tag does not match
+			}
+
+			if (device_number == reg_si) {
+				// Found!
+				reg_bl = static_cast<uint8_t>(i & 0xff);
+				reg_bh = 0x00; // bus 0
+				set_return_code(ReturnCode::Successful);
+				return;
+			}
+
+			// Matches, but is not the SIth device
+			++device_number;
+		}
+
+		// Device not found
+		set_return_code(ReturnCode::DeviceNotFound);
+		return;
+	}
+	case 0x03: { // Find PCI Class Code
+		const auto tmp = reg_ecx & 0xffffff;
+		const uint32_t class_tag = static_cast<uint32_t>(tmp);
+
+		// Try to find requested device
+		uint32_t device_number = 0;
+		for (uint16_t i = 0; i <= max_device_index; ++i) {
+			// Query unique device/subdevice entries
+			const auto i_shifted = static_cast<uint32_t>(i << 8);
+			IO_WriteD(port_num_pci_config_address,
+			          static_cast<uint32_t>(0x80000000 | i_shifted));
```

Comment:
`0x80000000` is used in quite a few places. I'm guessing it's some kind of base address. Could we extract it to a constant?

Reply:
If I'm not mistaken, this is _enable bit_ (see https://wiki.osdev.org/PCI). Named accordingly, converted to `1 << 31` to be more clear what it is.